lib/t/feeds.ex

```diff
@@ -115,42 +138,97 @@ defmodule T.Feeds do
+         location,
+         gender,
+         gender_preferences,
+         min_age,
+         max_age,
+         distance,
+         count
+       ) do
     feeded = FeededProfile |> where(for_user_id: ^user_id) |> select([s], s.user_id)
 
     most_liked_count = count - div(count, 2)
 
     most_liked =
-      feed_profiles_q(user_id, gender, gender_preferences)
-      |> where([p], p.user_id not in subquery(feeded))
-      |> order_by(desc: :times_liked)
+      most_liked_q(user_id, gender, gender_preferences, feeded)
+      |> maybe_apply_age_filters(min_age, max_age)
+      |> maybe_apply_distance_filter(location, distance)
       |> limit(^most_liked_count)
-      |> select([p], {p, distance_km(^location, p.location)})
       |> Repo.all()
 
-    filter_out_ids = Enum.map(most_liked, fn {p, _} -> p.user_id end)
+    filter_out_ids = Enum.map(most_liked, fn p -> p.user_id end)
 
     most_recent_count = count - length(most_liked)
 
     most_recent =
-      feed_profiles_q(user_id, gender, gender_preferences)
-      |> where([p], p.user_id not in subquery(feeded))
-      |> where([p], p.user_id not in ^filter_out_ids)
-      |> order_by(desc: :last_active)
+      most_recent_q(user_id, gender, gender_preferences, feeded, filter_out_ids)
+      |> maybe_apply_age_filters(min_age, max_age)
+      |> maybe_apply_distance_filter(location, distance)
       |> limit(^most_recent_count)
-      |> select([p], {p, distance_km(^location, p.location)})
       |> Repo.all()
 
     most_liked ++ most_recent
   end
 
+  defp most_liked_q(user_id, gender, gender_preferences, feeded) do
+    feed_profiles_q(user_id, gender, gender_preferences)
+    |> where([p], p.user_id not in subquery(feeded))
+    |> order_by(desc: :times_liked)
+  end
+
+  defp most_recent_q(user_id, gender, gender_preferences, feeded, filter_out_ids) do
+    feed_profiles_q(user_id, gender, gender_preferences)
+    |> where([p], p.user_id not in subquery(feeded))
+    |> where([p], p.user_id not in ^filter_out_ids)
+    |> order_by(desc: :last_active)
+  end
+
+  defp maybe_apply_age_filters(query, min_age, max_age) do
+    query
+    |> maybe_apply_min_age_filer(min_age)
+    |> maybe_apply_max_age_filer(max_age)
+  end
+
+  defp maybe_apply_min_age_filer(query, min_age) do
+    if min_age do
+      %{year: y, month: m, day: d} = DateTime.utc_now()
+      youngest = %Date{year: y - String.to_integer(min_age), month: m, day: d}
+
+      where(query, [p], p.birthdate <= ^youngest)
+    else
+      query
+    end
+  end
+
+  defp maybe_apply_max_age_filer(query, max_age) do
+    if max_age do
+      %{year: y, month: m, day: d} = DateTime.utc_now()
+      oldest = %Date{year: y - String.to_integer(max_age), month: m, day: d}
+
+      where(query, [p], p.birthdate >= ^oldest)
+    else
+      query
+    end
+  end
+
+  defp maybe_apply_distance_filter(query, location, distance) do
+    if distance do
+      where(query, [p], distance_km(^location, p.location) <= ^distance)
```

Comment:
btw this is too slow since it can't use any index

Reply:
https://github.com/bryanjos/geo_postgis/blob/80ebb06b03578ec0e170de31ce734674ca79cb69/lib/geo_postgis.ex#L62-L70

Reply:
so `st_dwithin_in_meters` seems to be fine

Reply:
more info https://postgis.net/docs/ST_DWithin.html

Reply:
any easy fix in your mind?

Reply:
@ruslandoga  to clarify - current code is fine in the end of the day because it already uses `st_dwithin_in_meters`?

Reply:
no?